include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -5705,37 +5705,50 @@ namespace internal
   {
     template <int fe_degree, int = 0>
     static bool
-    run(const unsigned int           n_desired_components,
-        const AlignedVector<Number> &inverse_shape,
-        const AlignedVector<Number> &inverse_coefficients,
-        const Number *               in_array,
-        Number *                     out_array,
+    run(const unsigned int                          n_desired_components,
```

Comment:
You can write the code like this:
```cpp
    template <int fe_degree, int = 0>
    static bool
    run(const unsigned int                          n_desired_components,
        const FEEvaluationData<dim, Number, false> &fe_eval,
        const AlignedVector<Number> &               inverse_coefficients,
        const bool                                  dyadic_coefficients,
        const Number *                              in_array,
        Number *                                    out_array)
    {
      const auto dofs_per_component = fe_degree == -1 ? 
         fe_eval.get_shape_info().dofs_per_component_on_cell  : 
         Utilities::pow(fe_degree + 1, dim);
```
This allows to skip the second implementation.

Reply:
In which sense it is different? You mean these two:

https://github.com/dealii/dealii/blob/4fed85d2113ecfe45d27e7189d6c290f797985e3/include/deal.II/matrix_free/evaluation_kernels.h#L5742-L5746

https://github.com/dealii/dealii/blob/4fed85d2113ecfe45d27e7189d6c290f797985e3/include/deal.II/matrix_free/evaluation_kernels.h#L5847

Why should eo not work for `-1`?

Reply:
Yes, that's what I meant. I don't know about the `EvaluateTensorProduct` at all. I only tried to judge by the existing implementation in `Basic`... I assumed maybe it needed `fe_degree` not `-1`, or smt like that. But if you say, it would/could work, I'll of course try it!

Reply:
It should work. You never plug in `fe_degree` but `fe_degree+1`, which `0` for the `-1` case as in the second code snippet!

Reply:
You are right 👍  I feared `<... 0, 0, ...>` would only work for `evaluate_general`, which was a false inference from too little data from my side 🤣. Sorry about that. I guess the existence of this template specialization also confirms it. https://github.com/dealii/dealii/blob/947199be8142d3fc7bde15d8f94f168573ccf5d2/include/deal.II/matrix_free/tensor_product_kernels.h#L1987-L1994

Reply:
Just to answer your question above. This template specialization is rather new (see https://github.com/dealii/dealii/commit/8e6e6eb08137e98295ad874aa59028b63711e4f3); we added it for a benchmark: https://github.com/peterrum/dealii-dd-and-schwarz/blob/677fcbfb25bda8b24e8cd7b842cf9bff1652e9b9/include/operator.h#L1245. I guess we can remove multiple template specialization now in the [dealii/include/deal.II/matrix_free/evaluation_kernels.h](https://github.com/dealii/dealii/blob/4fed85d2113ecfe45d27e7189d6c290f797985e3/include/deal.II/matrix_free/evaluation_kernels.h#L5742-L5746) file, since they are from a time when such a specialization did not exist!

Reply:
So do you mean that there should be only one `run()` which uses `evaluate_evenodd` evaluator, i.e.: 
```c++
const unsigned int given_degree =
        (fe_degree > -1) ? fe_degree :
                           fe_eval.get_shape_info().data.front().fe_degree;

// ...

EvaluatorTensorProduct<evaluate_evenodd, dim, 0, 0, Number> evaluator(
        AlignedVector<Number>(),
        AlignedVector<Number>(),
        fe_eval.get_shape_info().data.front().inverse_shape_values_eo,
        given_degree + 1,
        given_degree + 1);
```
This will allow me to skip the second implementation. However, in light of  https://github.com/dealii/dealii/pull/14850#discussion_r1133073361, I would still have two branches inside this single `run()` to differentiate between scalar and dyadic coefficients.

The coupling nature of the dyadic coefficients require that the second time the shape functions are evaluated happens in a separate loop after the coefficients are multiplied. This means that I cannot hide away the coefficient step in a `vmult` (what you mentioned in https://github.com/dealii/dealii/pull/14850#discussion_r1133044473). I will still implement it and call it as `vmult<dim>` or `vmult<-1>` hoping the `dim` case will bring some performance. Or am I wrong? This could look like:
```c++
if (n_desired_components == dim)
                vmult<dim>(
                  inv_coefficient, tmp, out_array, dofs_per_component, q);
              else
                vmult<-1>(inv_coefficient,
                          tmp,
                          out_array,
                          dofs_per_component,
                          q,
                          n_desired_components);
```
Sorry about bringing all the issues together in this comment but they are all related 😆

Reply:
First part:
```cpp
const unsigned int given_degree =
        (fe_degree > -1) ? fe_degree :
                           fe_eval.get_shape_info().data.front().fe_degree;

// ...

EvaluatorTensorProduct<evaluate_evenodd, dim, fe_degree + 1, fe_degree + 1, Number> evaluator(
        AlignedVector<Number>(),
        AlignedVector<Number>(),
        fe_eval.get_shape_info().data.front().inverse_shape_values_eo,
        given_degree + 1,
        given_degree + 1);
```
Second part: 
> hoping the dim case will bring some performance. Or am I wrong? This could look like:

Yes it is about performance. Having some constant loop bounds would be nice. Are there typical values for `n_desired_components`, which make sense to template with code as you suggested? `dim` was just a suggestion, as I thought that this is a typical value in the context of darcy flow.

Reply:
I'd be perfectly happy with `dim` for Darcy flow and `1` is also an obvious choice. Judging by the `MassOperator` in ExaDG ([here](https://github.com/exadg/exadg/blob/c71a5bf0971e5463008158e4643bde7a775f1009/include/exadg/operators/mass_operator.cpp#L86-L112)), it might be a good idea to provide `dim+1` and `dim+2` as well. I can do that :+1:.

Reply:
Let me address two of the comments:
> I guess we can remove multiple template specialization now in the [dealii/include/deal.II/matrix_free/evaluation_kernels.h](https://github.com/dealii/dealii/blob/4fed85d2113ecfe45d27e7189d6c290f797985e3/include/deal.II/matrix_free/evaluation_kernels.h#L5742-L5746) file, since they are from a time when such a specialization did not exist!

We should measure this in terms of instructions spent inside these functions, say for degree 1 and degree 3 in 2D & 3D, or alternatively by assessing the generated assembly code. I am all in favor of the present PR, but this comment is about an unrelated topic and should be kept separate. The reason is that I consider it a high priority of avoiding mistakes done in the past where we accumulated too much overhead in the call functions by seemingly unrelated functions. I recall how much work it was to do #13056 (described much earlier in #10927 and #9794), so we should make such bigger changes by assessing the number of instructions spent and/or inspecting the assembly code. The background is that we spend **way too many** instructions for setting up the right call interfaces for low polynomial degrees `p=1,2` or in 2D, which is a real problem.

> I'd be perfectly happy with dim for Darcy flow and 1 is also an obvious choice.

I don't think we need more than these two cases for the tensor-valued case. I cannot imagine what kind of coupling would be an actual underlying physics problem in that scenario. Of course, you might want to have a coupled system with more variables, but not the dense tensorial coupling. In such a case, it be better to have multiple evaluators and we should rather focus on reducing the individual overhead (see previous paragraph) instead.